### ISSUE TITLE
Set up internationalization (frontend)

### ISSUE THREAD

**Issue by sandalwoodbox** [FanvidDB/fanviddb-server#3]
Setting up internationalization at the beginning of a project is so much easier than trying to do it later on.

This ticket is for setting up extraction of strings from the frontend.

**Comment by sandalwoodbox** [comment]
Here are some notes about what I've found so far. There doesn't seem to be a single winner in this area for javascript.

- <react.i18next.com> seems to just be trying to sell me on a product. How to actually use it is unclear.
- <formatjs.io> seems like it's pretty straightforward to use, but they expect you to be working with a localization service so they don't have an official mechanism for using the extracted messages to set up translation files. All you can get off the bat is basically a template file (in the parlance of [babel (python)](http://babel.pocoo.org/en/latest/messages.html#message-extraction).
  - [react-intl-translations-manager](https://github.com/GertjanReynaert/react-intl-translations-manager/) is pretty slick but hasn't been updated for 3 years
  - This does have integrations with React that seem potentially useful?
- [polyglot.js](https://github.com/airbnb/polyglot.js) seems fine in terms of marking things for translation & getting translations from a dict, but also doesn't have any mechanism for managing translations.
- The babel python package does also support extracting simple translations from Javascript / HTML / custom parsers and adding them to `.po` files; that should be fine. There's maybe a slight concern about page weight from non-JS translations that would get stuck in with the JS translations, but tbh it's probably not much; most of the translation should be on the JS side, and overall there shouldn't be that many things translated. This could likely be used in combination with any of the other options.

**Comment by sandalwoodbox** [comment]
Just to make it explicit: None of the JS packages for this seem to have any special handling of translations; ultimately they all just expect to load translations from an object that maps a "key" to a translation for a specific locale. Where that translation is coming from is undefined and there are basically no recommendations.

**Comment by sandalwoodbox** [comment]
It looks like there are also some libraries out there specifically for providing gettext support in Javascript:

- [gettext.js](https://github.com/Orange-OpenSource/gettext.js)
- [Jed](http://messageformat.github.io/Jed/)
- [po2json](https://github.com/mikeedwards/po2json) - turns po files to Jed-compatible JSON.

These also seem to be rarely updated (though that may just be because they are already basically "complete".

There is also [messageformat](http://messageformat.github.io/) which seems to have a different opinion of how to handle translations, but also has some amount of compatibility with gettext. It does also have a react integration, and is a project of the OpenJS Foundation.

**Comment by sandalwoodbox** [comment]
Some comparisons of gettext vs messageformat:

- https://stackoverflow.com/questions/13734622/is-it-a-good-idea-to-use-gettext-and-icu-messageformat-combined

[Slack uses messageformat](https://slack.engineering/localizing-slack/) - they chose it partially due to good php support (python projects tend to go for gettext because of the builtin support & the babel library, it seems).

**Comment by sandalwoodbox** [comment]
It looks like [gettext can extract messageformat strings](https://www.gnu.org/software/gettext/manual/gettext.html#java_002dformat) - assuming I'm reading the docs right 😂  Nobody mentioned this otherwise.

**Comment by sandalwoodbox** [comment]
It does look like [formatjs and i18next are much more popular than jed / gettext](https://www.npmtrends.com/gettext-vs-i18n-vs-i18next-vs-intl-messageformat-vs-jed) though popularity of all of them has gone up over time. (i18n is only nodejs / server-side so it's not relevant.)

**Comment by sandalwoodbox** [comment]
There is also a project called "Fluent" that [might be used by Firefox](https://stackoverflow.com/a/55387743) but its official [comparison to gettext](https://github.com/projectfluent/fluent/wiki/Fluent-vs-gettext) seems incorrect on several basic points to me.

**Comment by sandalwoodbox** [comment]
The [ICU guide to internationalization](https://unicode-org.github.io/icu/userguide/icu/i18n.html) is a good read in terms of getting a handle on what i18n can entail.

As I'm reading these docs.... I like the idea that message ids mean that content and code can be completely separate. In other words, the interface for people to contribute non-English copy could be the same as for people to contribute English copy. This would open up more opportunities for people to contribute.

**Comment by sandalwoodbox** [comment]
It seems like the MessageFormat libraries seem to favor stuffing the translation into a single string, which feels like it could get unwieldy compared to doing switches with gettext "context".

**Comment by sandalwoodbox** [comment]
Fluent does support [React](https://github.com/projectfluent/fluent.js/) and [Python](https://github.com/projectfluent/python-fluent) but explicitly [doesn't support extraction](https://github.com/projectfluent/fluent.js/wiki/React-Tutorial#aside-on-automatic-string-extraction). We'd need some way to detect and flag missing translations, which could probably only happen at runtime (because fluent doesn't require every message id to be in every file, due to their fallback system.)

It does seem like Fluent goes to an effort to make editing translations easy for non-programmers. The [Fluent Syntax Guide](https://www.projectfluent.org/fluent/guide/) seems pretty slick.

**Comment by sandalwoodbox** [comment]
Doc going through the different options: https://docs.google.com/document/d/1KwvQtD7mRkQDdBLRw2GECWgYd-xPzatBJ_ji8UNxgP8/edit

TLDR for our purposes fluent seems like the best option. It seems to offer the simplest and most understandable / expressive syntax for translators, which is important because we don't have access to a translation management system.

**Comment by sandalwoodbox** [comment]
Resolved as of 8079922.
